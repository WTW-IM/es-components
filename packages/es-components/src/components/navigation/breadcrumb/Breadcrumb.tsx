import React from 'react';
import type * as CSS from 'csstype';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import tinycolor from 'tinycolor2';
import useUniqueId from '../../util/useUniqueId';

function darkenPrimary(primary: CSS.Property.Color) {
  return tinycolor(primary).darken(15).toRgbString();
}

const OrderedList = styled.ol`
  background-color: ${props => props.theme.colors.gray1};
  border-radius: 2px;
  line-height: ${props => props.theme.font.baseLineHeight};
  list-style: none;
  margin-bottom: 25px;
  margin-top: 0;
  padding: 8px 15px;

  > li + li {
    padding-left: 24px;

    &::before {
      border-color: ${props => props.theme.colors.gray5};
      border-style: solid;
      border-width: 0 2px 2px 0;
      box-sizing: border-box;
      content: '';
      display: block;
      height: 13px;
      left: 0px;
      position: absolute;
      top: 20%;
      transform: rotate(-45deg);
      width: 13px;
    }
  }
`;

const Crumb = styled.li`
  color: ${props => props.theme.brandColors.primary3};
  display: inline-block;
  margin-right: 2px;
  position: relative;

  > a {
    color: ${props => props.theme.colors.primary};
    text-decoration: none;
  }

  > a:hover {
    color: ${props => darkenPrimary(props.theme.colors.primary)};
    text-decoration: underline;
  }
`;

const getChildKey = (child: React.ReactNode) =>
  (
    (child as Maybe<React.ReactElement | React.ReactPortal>)?.key ||
    (typeof child === 'string' ||
    typeof child === 'number' ||
    typeof child === 'boolean'
      ? child
      : undefined)
  )?.toString();

export type BreadcrumbProps = Override<
  JSXElementProps<'ol'>,
  {
    children: React.ReactNode;
  }
>;

function Breadcrumb({ children, ...props }: BreadcrumbProps) {
  const keyId = useUniqueId();
  return (
    <OrderedList {...props}>
      {React.Children.map(children, (child, index) => (
        <Crumb key={`${keyId}-${getChildKey(child)}-${index}`}>{child}</Crumb>
      ))}
    </OrderedList>
  );
}

Breadcrumb.propTypes = {
  children: PropTypes.node.isRequired
};

export default Breadcrumb;
