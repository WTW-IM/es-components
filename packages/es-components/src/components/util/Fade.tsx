import React from 'react';
import PropTypes from 'prop-types';
import Transition, {
  TransitionStatus,
  TimeoutProps
} from 'react-transition-group/Transition';

type Opacity = React.CSSProperties['opacity'];

interface FadeProps extends Partial<TimeoutProps<HTMLElement>> {
  duration?: number;
  opacity?: Opacity;
  children: NonNullable<React.ReactNode>;
}

const getFadeStyle = (status: TransitionStatus, finalOpacity: Opacity) => {
  switch (status) {
    case 'entering':
    case 'entered':
      return { finalOpacity };
    default:
      return { opacity: 0 };
  }
};

type AnyReactHTMLElement = React.ReactHTMLElement<HTMLElement>;

const TransitionChild: React.FC<{
  children: AnyReactHTMLElement;
  style: React.CSSProperties;
  transitionStatus: TransitionStatus;
}> = ({ children, style, transitionStatus }) => {
  return React.cloneElement(children, {
    style: {
      ...((children?.props?.style as Maybe<React.CSSProperties>) || {}),
      ...style,
      ...getFadeStyle(
        transitionStatus,
        typeof style.opacity == 'number' || Boolean(style.opacity)
          ? style.opacity
          : 1
      )
    }
  });
};

const Fade: React.FC<FadeProps> = ({
  children,
  duration,
  opacity,
  ...otherProps
}) => {
  const transitionStyles = {
    transition: `opacity ${duration || 150}ms linear`,
    opacity
  };

  const transitionProps = {
    ...otherProps,
    timeout: duration || otherProps.timeout || 150
  };

  const hasValidChildren =
    children &&
    (typeof children === 'object' || typeof children === 'function');

  return !hasValidChildren ? (
    <>{children}</>
  ) : (
    <Transition {...transitionProps}>
      {transitionStatus => {
        try {
          return (
            React.Children.only(children) && (
              <TransitionChild
                transitionStatus={transitionStatus}
                style={transitionStyles}
              >
                {children as AnyReactHTMLElement}
              </TransitionChild>
            )
          );
        } catch (e) {
          return React.Children.map(children, child => {
            return !child ? (
              child
            ) : (
              <TransitionChild
                transitionStatus={transitionStatus}
                style={transitionStyles}
              >
                {child as AnyReactHTMLElement}
              </TransitionChild>
            );
          });
        }
      }}
    </Transition>
  );
};

Fade.propTypes = {
  children: PropTypes.any.isRequired,
  duration: PropTypes.number,
  opacity: PropTypes.number
};

Fade.defaultProps = {
  duration: 150,
  opacity: 1
};

export default Fade;
