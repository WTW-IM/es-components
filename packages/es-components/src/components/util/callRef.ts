import React from 'react';

const callRef = <T, R extends React.ForwardedRef<T>>(ref: Maybe<R>, value: T) =>
  typeof ref === 'function'
    ? ref(value)
    : ref && Object.hasOwn(ref, 'current') && (ref.current = value);

export default callRef;

export function callRefs<T, R extends React.ForwardedRef<T>>(
  value: T,
  ...refs: R[]
) {
  refs.forEach(r => callRef(r, value));
}
