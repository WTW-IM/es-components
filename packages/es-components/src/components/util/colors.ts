import tinycolor, { ColorInput } from 'tinycolor2';

export function darken(color: ColorInput, percent: number) {
  return tinycolor(color).darken(percent).toHexString();
}

export function lighten(color: ColorInput, percent: number) {
  return tinycolor(color).lighten(percent).toHexString();
}

export function getTextColor(
  bgColor: ColorInput,
  darkestColor: ColorInput = 'black',
  lightestColor: ColorInput = 'white'
) {
  return tinycolor.readability(darkestColor, bgColor) > 10
    ? darkestColor
    : lightestColor;
}
