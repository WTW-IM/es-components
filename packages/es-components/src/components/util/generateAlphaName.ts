export default function generateAlphaName(): string {
  const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
  let text = '';
  for (let i = 0; i < 6; i++) {
    text += possible.charAt(Math.floor(Math.random() * possible.length));
  }
  return text;
}
