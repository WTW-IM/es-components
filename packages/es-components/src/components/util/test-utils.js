// eslint-env jest
import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render } from '@testing-library/react';

export function ThemeComponent(props) {
  return <ThemeProvider theme={viaTheme} {...props} />;
}

ThemeComponent.propTypes = {
  children: PropTypes.node
};

export function renderWithTheme(component) {
  return render(<ThemeComponent>{component}</ThemeComponent>);
}
