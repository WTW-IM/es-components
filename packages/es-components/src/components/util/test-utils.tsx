import React, { Profiler } from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render, RenderResult } from '@testing-library/react';

export function ThemeComponent(
  props: Omit<React.ComponentPropsWithoutRef<typeof ThemeProvider>, 'theme'>
) {
  return <ThemeProvider {...props} theme={viaTheme} />;
}

ThemeComponent.propTypes = {
  children: PropTypes.node
};

const profilerRender = (id: string, phase: string, actualDuration: number) => {
  if (actualDuration < 100) return;
  console.warn('Render took too long!', { id, phase, actualDuration });
};

export function renderWithTheme(component: React.ReactElement): RenderResult {
  const { ...viewUtils } = render(<ThemeComponent>{component}</ThemeComponent>);
  return {
    ...viewUtils,
    rerender: (rerenderComponent: React.ReactNode) =>
      viewUtils.rerender(
        <Profiler id="ThemeRender" onRender={profilerRender}>
          <ThemeComponent>{rerenderComponent}</ThemeComponent>
        </Profiler>
      )
  };
}

export const setClientWidth = (width: number) => {
  Object.defineProperty(document.documentElement, 'clientWidth', {
    writable: true,
    configurable: true,
    value: width
  });
};

export const setClientHeight = (height: number) => {
  Object.defineProperty(document.documentElement, 'clientHeight', {
    writable: true,
    configurable: true,
    value: height
  });
};
