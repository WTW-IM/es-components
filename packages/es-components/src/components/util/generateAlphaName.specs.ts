import generateAlphaName from './generateAlphaName';

describe('generateAlphaName', () => {
  it('generates a 6 character string', () => {
    const result = generateAlphaName();

    expect(result.length).toBe(6);
  });

  it('generates only alpha characters', () => {
    const result = generateAlphaName();

    expect(/^[a-zA-Z]+$/.test(result)).toBe(true);
  });

  it('generates a unique string each subsequent execution', () => {
    const result1 = generateAlphaName();
    const result2 = generateAlphaName();

    expect(result1).not.toEqual(result2);
  });
});
