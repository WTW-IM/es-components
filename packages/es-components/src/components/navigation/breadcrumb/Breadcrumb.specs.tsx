import React from 'react';
import Breadcrumb from './Breadcrumb';
import { renderWithTheme } from '../../util/test-utils';
import { screen } from '@testing-library/react';

it('renders as expected', () => {
  const { container } = renderWithTheme(
    <Breadcrumb>
      <a href="/test">test</a>
      <a href="/test-2">test 2</a>
      <a href="/test-3">test 3</a>
    </Breadcrumb>
  );

  expect(container).toMatchSnapshot();
});

it('handles updates in fragment children', async () => {
  const { rerender } = renderWithTheme(
    <Breadcrumb>
      <a href="/test">test</a>
      <a href="/test-2">test 2</a>
      <>
        <a href="/test-3">test 3</a>
        <span>hello</span>
      </>
    </Breadcrumb>
  );

  expect(await screen.findByText('hello')).toBeInTheDocument();

  rerender(
    <Breadcrumb>
      <a href="/test">test</a>
      <a href="/test-2">test 2</a>
      <>
        <a href="/test-3">test 3</a>
        <span>hi</span>
      </>
    </Breadcrumb>
  );
  expect(await screen.findByText('hi')).toBeInTheDocument();
});
