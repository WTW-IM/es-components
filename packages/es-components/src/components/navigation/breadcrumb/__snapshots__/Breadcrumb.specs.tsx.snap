// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
.c0 {
  background-color: #f4f4f4;
  border-radius: 2px;
  line-height: 1.428;
  list-style: none;
  margin-bottom: 25px;
  margin-top: 0;
  padding: 8px 15px;
}

.c0 > li + li {
  padding-left: 24px;
}

.c0 > li + li::before {
  border-color: #c5c5c5;
  border-style: solid;
  border-width: 0 2px 2px 0;
  box-sizing: border-box;
  content: '';
  display: block;
  height: 13px;
  left: 0px;
  position: absolute;
  top: 20%;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  width: 13px;
}

.c1 {
  color: #c110a0;
  display: inline-block;
  margin-right: 2px;
  position: relative;
}

.c1 > a {
  color: #0073b6;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 > a:hover {
  color: rgb(0,67,105);
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div>
  <ol
    class="c0"
  >
    <li
      class="c1"
    >
      <a
        href="/test"
      >
        test
      </a>
    </li>
    <li
      class="c1"
    >
      <a
        href="/test-2"
      >
        test 2
      </a>
    </li>
    <li
      class="c1"
    >
      <a
        href="/test-3"
      >
        test 3
      </a>
    </li>
  </ol>
</div>
`;
